File: DSA/Manav-Rajput/Day44_question2.java
class Solution {
    public int maxSumTwoNoOverlap(int[] nums, int firstLen, int secondLen) {
        return Math.max(maxSum(nums, firstLen, secondLen), maxSum(nums, secondLen, firstLen));
    }
    
    private int maxSum(int[] nums, int firstLen, int secondLen) {
        int n = nums.length;
        int[] prefixSum = new int[n + 1];
        for (int i = 0; i < n; i++) {
            prefixSum[i + 1] = prefixSum[i] + nums[i];
        }
        int maxFirstLen = 0;
        int result = 0;
        
        for (int i = firstLen + secondLen; i <= n; i++) {
            maxFirstLen = Math.max(maxFirstLen, prefixSum[i - secondLen] - prefixSum[i - secondLen - firstLen]);
            result = Math.max(result, maxFirstLen + (prefixSum[i] - prefixSum[i - secondLen]));
        }
        return result;
    }
}

File: DSA/Arshdeep-Kaur/Day 44/Day44_Question2.cpp
class Solution {
public:
    int maxSumTwoNoOverlap(vector<int>& nums, int firstLen, int secondLen) {
        int n = nums.size();
        vector<int> arr(n + 1, 0);

        //calculate the prefix sums
        for(int i=0; i<n; i++){
            arr[i+1] = nums[i] + arr[i];
        }

        int sum=0;

        // First try: firstLen subarray comes before secondLen subarray
        for(int i=0; i<=n-firstLen-secondLen; i++){
            int firstsum = arr[i+firstLen] - arr[i];
            for(int j=i+firstLen; j<=n-secondLen; j++){
                int secondsum = arr[j+secondLen] - arr[j];
                sum = max(sum,firstsum+secondsum);
            }
        }

        // Second try: secondLen subarray comes before firstLen subarray
        for(int i=0; i<=n-firstLen-secondLen; i++){
            int secondsum = arr[i+secondLen] - arr[i];
            for(int j=i+secondLen; j<=n-firstLen; j++){
                int firstsum = arr[j+firstLen] - arr[j];
                sum = max(sum,firstsum+secondsum);
            }
        }
        
        return sum;
    }
};
